File: po_lab/szybkosc_javy/largest_clique_size.cpp
#include <iostream>
#include <vector>
#include <cstdint>
#include <bit>
#include <algorithm>
using namespace std;

class LargestCliqueSize
{
private:
    int n, m;
    vector<vector<bool>> adjacent;

    void input()
    {
        cin >> n >> m;
        adjacent = vector<vector<bool>>(n, vector<bool>(n));

        for (int i = 0; i < m; i++)
        {
            int u, v;
            cin >> u >> v;
            adjacent[u][v] = true;
            adjacent[v][u] = true;
        }
    }

    bool check_mask(uint64_t mask)
    {
        for (int i = 0; i < n; i++)
        {
            if (!(mask & (1ULL << i))) continue;
            
            for (int j = i + 1; j < n; j++)
            {
                if (!(mask & (1ULL << j))) continue;
                if (!adjacent[i][j]) return false;
            }
        }

        return true;
    }

public:
    void find_largest_clique_size()
    {
        input();
        int result = 0;

        for (uint64_t mask = 0; mask < 1ULL << n; mask++)
        {
            if (check_mask(mask)) result = max(result, popcount(mask));
        }

        cout << result << "\n";
    }
};

int main()
{
    LargestCliqueSize counter{};
    counter.find_largest_clique_size();
}

File: po_lab/szybkosc_javy/LargestCliqueSize.java
import java.util.Scanner;

public class LargestCliqueSize {
    private static int n, m;
    private static boolean[][] adjacent;

    private static void input()
    {
        Scanner scanner = new Scanner(System.in);
        n = scanner.nextInt();
        m = scanner.nextInt();
        adjacent = new boolean[n][n];

        for (int i = 0; i < m; i++)
        {
            int u, v;
            u = scanner.nextInt();
            v = scanner.nextInt();
            adjacent[u][v] = true;
            adjacent[v][u] = true;
        }
    }

    private static boolean check_mask(long mask)
    {
        for (int i = 0; i < n; i++)
        {
            if ((mask & (1L << i)) == 0) continue;
            
            for (int j = i + 1; j < n; j++)
            {
                if ((mask & (1L << j)) == 0) continue;
                if (!adjacent[i][j]) return false;
            }
        }

        return true;
    }

    private static void find_largest_clique_size()
    {
        input();
        int result = 0;

        for (long mask = 0; mask < 1L << n; mask++)
        {
            if (check_mask(mask)) result = Math.max(result, Long.bitCount(mask));
        }

        System.out.println(result);
    }

    public static void main(String[] args)
    {
        find_largest_clique_size();
    }
}
